1.sort_namesメソッドの処理内容を変更

Ryooo-k opened the pull request:
2.-lオプションの追加
3.try_load_attributesメソッドで詳細情報を取得
4.変数名、メソッド名を変更

Ryooo-k committed:
1.sort_namesメソッドの処理内容を変更
2.-lオプションの追加
3.try_load_attributesメソッドで詳細情報を取得
4.変数名、メソッド名を変更

Ryooo-k committed:
1.メソッド外でオプション可否による実行分岐を設定
2.1の変更に伴うメソッド名・変数名の変更

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,43 +1,135 @@
 COLUMN = 3
-SPACE = 2
+FORMAT_TIME = '%b %d %H:%M'
+BLOCK_SIZA = 1024
```
```suggestion
BLOCK_SIZE = 1024
```

かな？

スペルチェッカーを入れておくといいかもしれません。

[Code Spell Checker でスペルミスをチェックする \[VSCode\] \- もとひろ blog](https://motohiro-mm.hatenablog.com/entry/2023/10/11/131354)

Ryo-ta090972 replied:
ありがとうございます🙇‍♂️
スペルチェッカー導入します！

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,43 +1,135 @@
 
 COLUMN = 3
-SPACE = 2
+FORMAT_TIME = '%b %d %H:%M'
```
「時間をフォーマットする」と読めてしまうので、`TIME_FORMAT`とした方が自然ですね。
ただ、`strftime`メソッドの引数は定数にしなくてもベタ書きで構わないと思います。

```ruby
strftime('%b %d %H:%M')
```

Ryo-ta090972 replied:
これはベタ書きでも良いのですね😯
ご指摘ありがとうございます！

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,43 +1,135 @@
+  names = Dir.entries(target_dir_path)
+  sorted_names = sort_names(names, options)
+  filtered_names = filter_names(sorted_names, options)
+  formated_names = options[:l] ? format_names(load_names_attribute(filtered_names, target_dir_path), options) : format_names(filtered_names, options)
```
lオプションあり・なしで表示ロジックが大きく異なるので、`format_names`で全部まかなおうとせずに、メソッドを別に分けた方がいいと思います。

Ryo-ta090972 replied:
承知しました🙇‍♂️

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,43 +1,135 @@
+  names.each.map do |name|
+    name_path = File.absolute_path(name, path)
+    name_attribute = File::Stat.new(name_path)
+    total_block_size += name_attribute.blksize.to_i / BLOCK_SIZA
```
もともと整数値を返すのでto_iは不要ですね。
https://docs.ruby-lang.org/ja/latest/method/File=3a=3aStat/i/blksize.html

Ryo-ta090972 replied:
ご指摘ありがとうございます🙇‍♂️

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,43 +1,135 @@
-  repositioned_names.each.map do |names_for_row|
+  names.each.map do |name|
+    name_path = File.absolute_path(name, path)
+    name_attribute = File::Stat.new(name_path)
```
File::Statオブジェクトなので、`stat = `とか`file_stat = `とした方が自然かなと思います。

[ユーザーを見つけたら……Aさん「自転車に入れます」Bさん「人間に入れます」Cさん「ユーザーに入れます」 \| FBC](https://bootcamp.fjord.jp/pages/readable-variable-name)

Ryo-ta090972 replied:
ご指摘ありがとうございます！

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,43 +1,135 @@
+    '10' => '-',
+    '12' => 'l',
+    '14' => 's'
+  }[number]
```
毎回同じハッシュを使うのであれば定数化しましょう。

Ryo-ta090972 replied:
承知しました🙇‍♂️
定数化します🙇‍♂️

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,43 +1,135 @@
+    '4' => { 'x' => 's', '-' => 'S' }
+  }[number]
+
+  permission[index] = authority[permission[index]]
```
引数の内容を変えるメソッドはNGです〜。

https://bootcamp.fjord.jp/pages/do-not-use-destructive-method#-1

Ryo-ta090972 replied:
承知しました。
修正します。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,43 +1,135 @@
     names_for_row.each_with_index.map do |name, col|
-      name.to_s.ljust(max_str_sizes[col] + SPACE)
+      if options[:l] && !array_last?(names_for_row, col)
+        "#{name.to_s.rjust(max_str_sizes[col])} "
```
lオプションありのとき、（最後の項目以外）全部右寄せにするのは、はたして正しいでしょうか？
（データ的にたまたま問題が発覚しないだけなのでは？）

Ryo-ta090972 replied:
違うディレクトリで確認しましたが、オーナー名とグループ名は左寄せになっていました💦

```
 % ruby ls.rb -la ~
total 160
drwxr-xr-x 15  ryo  ryo  4096 Oct 28 23:31 .
drwxr-xr-x  3 root root  4096 Jun 08 10:29 ..
#中略

 % ls -la ~
total 336
drwxr-xr-x 15 ryo  ryo   4096 Oct 28 23:31 .
drwxr-xr-x  3 root root  4096 Jun  8 10:29 ..
#中略
```

また、（ついでにわかったことですが）totalの数値もあっていないため、こちらも合わせて修正いたします🙇‍♂️

JunichiIto reviewed (commented):
コメントしました〜。

Ryooo-k committed:
1.-lオプションの有無でそれぞれfomatメソッドを定義
2.-lオプションの有の時の表示形式を変更
3.strftimeの引数を定数からベタ書きに変更
4.スペルミスの修正
5.追加・変更がないハッシュを定数で定義
6.引数を変更するメソッドを変更しないメソッドに修正

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,57 +1,149 @@
+  names = Dir.entries(target_dir_path)
+  sorted_names = sort_names(names, options)
+  filtered_names = filter_names(sorted_names, options)
+  loaded_attributes = load_attributes(filtered_names, target_dir_path) if options[:l]
+  options[:l] ? puts(format_attributes(loaded_attributes)) : puts(format_names(filtered_names))
```
```suggestion
  options[:l] ? puts(format_attributes(filtered_names, target_dir_path)) : puts(format_names(filtered_names))
```

`options[:l]`でしか使わないデータはmainで読み込まなくていいと思います。

あと、参考演算子とputsが組み合わさると読みづらいので、いったん変数に入れる方がいいと思います。

```ruby
hoge = piyo ? ... : ...
puts hoge
```

Ryo-ta090972 replied:
承知しました。
修正します！

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,57 +1,149 @@
+    attributes << file_stat.size
+    attributes << file_stat.mtime.strftime('%b %e %H:%M')
+    attributes << name
+  end.prepend(["total #{total_block_size}"])
```
prependは「後から配列の先頭に突っ込む」メソッドなので、直感的にコードを読みづらくなります。

```ruby
hoge = piyo.map ...
[total, *hoge]
```

みたいな形でmapの戻り値をいったん変数に入れて、そこから新しく配列を作る方が読みやすいかなーと思います。

Ryo-ta090972 replied:
アドバイスありがとうございます！
修正します

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,57 +1,149 @@
+    attributes << load_type_and_permission(file_stat)
+    attributes << file_stat.nlink
+    attributes << Etc.getpwuid(file_stat.uid).name
+    attributes << Etc.getpwuid(file_stat.gid).name
```
`getpwuid`を2回呼んでますが、これであってますか？

Ryo-ta090972 replied:
Etc.getpwuid(file_stat.gid).nameは`getgrgid`メソッドを使うべきでした💦
ご指摘ありがとうございます！

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,57 +1,149 @@
+end
+
+def format_attributes(attributes)
+  max_str_sizes = find_max_str_sizes(attributes)
```
今のロジックだと、totalの桁数に合わせてパーミッションの列幅も広がっていってしまいそうです。
（以下のようなイメージ）

```ruby
total 1200000
-rw-r--r--    1 jnito staff    0 Oct  7 10:39 c.txt
-rw-r--r--    1 jnito staff    0 Oct  7 10:41 d.txt
```

Ryo-ta090972 replied:
全然気が付きませんでした😯
ご指摘ありがとうございます！

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,57 +1,149 @@
+  permission_number = file_number[3, 3]
+
+  permission = to_permission(permission_number)
+  IS_AUTHORITY[authority_number] ? FILE_TYPE[type_number] + to_authority(permission, authority_number) : FILE_TYPE[type_number] + permission
```
`FILE_TYPE[type_number] +`をDRYにしたいです

Ryo-ta090972 replied:
承知しました🙇‍♂️

JunichiIto reviewed (commented):
コメントしました〜。

Ryooo-k committed:
1.options[:l]でしか使わないデータをmainからメソッド内で読み込む
2.prependを使用しない処理に変更
3.getpwuid(file_stat.gid)⇒getgrgid(file_stat.gid)に変更
4.列幅を統一する際にtotalの幅は無視する
5.「FILE_TYPE[type_number] +」をDRY

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,57 +1,154 @@
+  target_dir_path = ARGV.empty? ? '.' : ARGV[0]
+  names = Dir.entries(target_dir_path)
+  sorted_names = sort_names(names, options)
+  loaded_names = load_names(sorted_names, target_dir_path, options)
```
僕の前回のコメントと修正の方針が異なっていますね。
この実装だとoptionsの内容によってload_namesが返すデータの形式（データ型）が異なります。
条件によってメソッドの戻り値の型が変わるとメソッドの利用者が驚く原因になるのでなるべく避けた方がいいです。

別の観点でアドバイスすると、`options[:l]`を使った条件分岐を1箇所に絞りましょう。
今の実装だと54行目と70行目の2箇所にあります。
条件分岐が1つになれば自ずと問題も解決されるんじゃないかなーと思います。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,57 +1,154 @@
+def load_attributes(names, path)
+  total_block_size = 0
+
+  loaded_attributes = names.each.map do |name|
```
eachは不要ですね。

あと、85行目にattributesがいて、それを格納する変数がloaded_attributesだとあまり違いが明確ではないですね。（両方attributesじゃん、ってなる）
違いが明確になる名前を検討してみてください。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,57 +1,154 @@
+    attributes << Etc.getgrgid(file_stat.gid).name
+    attributes << file_stat.size
+    attributes << file_stat.mtime.strftime('%b %e %H:%M')
+    attributes << name
```
```suggestion
    attributes << name
    attributes
```

`<<`の戻り値をブロックの戻り値にするのはちょっと意図が明確ではないので、配列そのものを明示的に最後に返す方がいいと思います。

そもそもの話として、配列リテラルを返す、という方法もあります。

```ruby
[
  load_type_and_permission(file_stat),
  # ...
  name
]
```

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,57 +1,154 @@
+  end
+
+  total = ["total #{total_block_size}"]
+  [total, *loaded_attributes]
```
totalも返すならメソッド名は `load_total_and_attributes`とかでしょうか。
ただ、メソッド名にandが入るのは、メソッドの責務が複数あるというあまり良くない兆候だったりします。。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,57 +1,154 @@
+end
+
+def format_attributes(attributes)
+  max_str_sizes = find_max_str_sizes(attributes[1..])
```
このメソッドだけをぱっと見たとき、メソッドの読み手は「なぜ`1..`のように最初の要素だけ除外してるんだろう？🤔」と疑問に思いそうです。（コンテキストの理解、つまりattributesがここに至るまでどう生成されて、どんなデータが入っているのか、ということをメソッドの呼び出しを遡って把握する必要が出てくる）

[僕が考える「良いコード」 \- give IT a try](https://blog.jnito.com/entry/2023/07/12/115733)

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,57 +1,154 @@
+def format_attributes(attributes)
+  max_str_sizes = find_max_str_sizes(attributes[1..])
+
+  attributes.each.map do |attributes_for_row|
```
このeachも不要です。
また、`attributes_for_row`を格納している変数が`attributes`なのは何か命名に問題がありそうな臭いがしますね 🤔

JunichiIto reviewed (commented):
コメントしました〜。

Ryooo-k committed:
1.options[:l]の条件分岐を１か所のみに変更
2.load_attributes内のtotal_block_sizeを別メソッドで定義

Ryooo-k committed:
121行目の余分なスペースを削除

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -61,4 +104,75 @@ def find_max_str_sizes(nested_names)
+end
+
+def to_permission(numbers)
+  numbers.each_char.map do |number|
+    PERMISSION[number]
+  end.join
```
```suggestion
  numbers.gsub(/./, PERMISSION)
```

だけでもいけるかも？

Ryo-ta090972 replied:
上記コードのみでOKでした😯
アドバイスありがとうございます！！

JunichiIto reviewed (commented):
コメントしました〜。

Ryo-ta090972 commented on 04.ls/ls.rb:
```diff
@@ -50,8 +50,8 @@ def main
-  loaded_names = load_names(sorted_names, target_dir_path, options)
-  formatted_names = options[:l] ? format_attributes(loaded_names) : format_names(loaded_names)
+  filtered_names = filter_names(sorted_names, options)
+  formatted_names = options[:l] ? format_attributes(filtered_names, target_dir_path) : format_names(filtered_names)
```
formatted_names_or_attributesに変更いたします🙇‍♂️

Ryo-ta090972 commented on 04.ls/ls.rb:
```diff
@@ -121,48 +163,16 @@ def to_authority(permission, number)
-      str_sizes[col] ||= []
-      str_sizes[col] << name.to_s.size
-    end
+  names.each do |name|
```
ご指摘ありがとうございます！

Ryo-ta090972 commented on 04.ls/ls.rb:
```diff
@@ -61,4 +104,75 @@ def find_max_str_sizes(nested_names)
+  permission_number = file_number[3, 3]
+
+  permission = to_permission(permission_number)
+  authority_permission = IS_AUTHORITY[authority_number] ? to_authority(permission, authority_number) : permission
```
authority_or_permissionに変更します🙇‍♂️

Ryo-ta090972 commented on 04.ls/ls.rb:
```diff
@@ -61,4 +104,75 @@ def find_max_str_sizes(nested_names)
+  end
+end
+
+def load_type_and_permission(file)
```
引数のfileは、File::Statクラスの為fileとしていました💦
file_statに変更します。

Ryooo-k committed:
1.formatted_names⇒formatted_names_or_attributesに変更
2.count_total_block_sizeメソッドの戻り値をIntegerに変更
3.count_total_block_sizeメソッド内のeachをsumに変更
4.IS_AUTHORITY⇒INDEX_AUTHORITY.key?に変更
5.authority_permission⇒authority_or_permissionに変更
6.INDEX_AUTHORITY⇒AUTHORITY_INDEXに変更
7.load_type_and_permissionの引数をfile⇒file_statに変更
8.to_permissionメソッドをgsub(/./, PERMISSION)の処理に変更

Ryooo-k committed:
1.find_max_str_sizesの引数nested_text⇒nested_namesに変更

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -51,8 +44,8 @@ def main
   filtered_names = filter_names(sorted_names, options)
-  formatted_names = options[:l] ? format_attributes(filtered_names, target_dir_path) : format_names(filtered_names)
-  puts formatted_names
+  formatted_names_or_attributes = options[:l] ? format_attributes(filtered_names, target_dir_path) : format_names(filtered_names)
```
`authority_or_permission`の方はちょっと抽象化するのが難しそうだな〜と思うのですが、こっちの変数名はもうちょっと抽象化の余地があるんじゃないかなと思います。
基本的に `_or_` や `_and_` でつなげるような変数名やメソッド名は「なんか臭う」名前なので極力登場させない方がいいです。

[名前重要！具体的過ぎるメソッド名の弊害と適切な抽象度 \| FJORD BOOT CAMP（フィヨルドブートキャンプ）](https://bootcamp.fjord.jp/articles/78)

Ryo-ta090972 replied:
再考します！

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -92,9 +85,9 @@ def reposition(names)
 end
 
-def find_max_str_sizes(nested_texts)
+def find_max_str_sizes(nested_names)
```
findなのかな？算出のイメージなのでcalcとかの方がいいかもしれません

Ryo-ta090972 replied:
findよりcalculateの方がしっくりきますね🤔
ご指摘ありがとうございます！

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -139,40 +132,34 @@ def load_attributes(names, path)
-    PERMISSION[number]
-  end.join
+def to_permission(number)
+  number.gsub(/./, PERMISSION)
```
あまり大した仕事をしてないのでメソッド化しなくてもいいかもしれませんね。
[メソッドを定義した方がいいケース、しない方がいいケース \| FBC](https://bootcamp.fjord.jp/pages/when-you-should-define-method)

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -139,40 +132,34 @@ def load_attributes(names, path)
 
-def load_type_and_permission(file)
-  file_number = file.mode.to_s(8).rjust(6, '0')
+def load_type_and_permission(file_stat)
```
loadは何かを読み込むイメージですね。
構築する=buildみたいな名前の方がいいかもしれません

Ryo-ta090972 replied:
構築する=buildは、思いつきませんでした💦
ありがとうございます！

JunichiIto reviewed (commented):
コメントしました〜。

Ryooo-k committed:
1.変数名変更（formatted_names_or_attributes⇒output_text）
2.メソッド名変更（find_max_str_sizes⇒calculate_max_str_sizes）
3.メソッド名変更（load_type_and_permission⇒build_type_and_permission）
4.メソッド削除（to_permission）

JunichiIto reviewed (approved):
確認しました。これでOKです〜 🙆‍♂️

Ryo-ta090972 commented:
ありがとうございました！
引き続きよろしくお願いします！
